core/src/main/scala/org/http4s/multipart/MultipartParser.scala

```diff
@@ -388,134 +387,186 @@ object MultipartParser {
-              Pull.done
-          case Some(_) =>
-            tailrecParts[F](boundary, values, l, r, limit)
+        //We can abuse reference equality here for efficiency
+        //Since `splitOrFinish` returns `empty` on a capped stream
+        //However, we must have at least one part, so `splitOrFinish` on this function
+        //Indicates an error
+        if (r == streamEmpty) {
+          Pull.raiseError(MalformedMessageBodyFailure("Cannot parse empty stream"))
+        } else {
+          tailrecParts[F](boundary, l, r, expectedBytes, limit)
         }
-    }.stream
+    }
+  }
 
   private def tailrecParts[F[_]: Sync](
       b: Boundary,
-      values: Array[Byte],
-      last: Stream[F, Byte],
-      next: Stream[F, Byte],
-      limit: Int
-  ): Pull[F, Part[F], Unit] =
-    parseToPartStreamed[F](last, limit).pull.echo *> splitHalf[F](values, next).flatMap {
-      case (l, r) =>
-        r.pull.uncons.flatMap {
-          case None =>
-            Pull.done
-          case Some(_) =>
-            tailrecParts[F](b, values, l, r, limit)
+      headerStream: Stream[F, Byte],
+      rest: Stream[F, Byte],
+      expectedBytes: Array[Byte],
+      limit: Int): Pull[F, Part[F], Unit] =
+    Pull
+      .eval(parseHeaders(headerStream))
+      .flatMap { hdrs =>
+        splitHalf(expectedBytes, rest).flatMap {
+          case (l, r) =>
+            //We hit a boundary, but the rest of the stream is empty
+            //and thus it's not a properly capped multipart body
+            if (r == streamEmpty) {
+              Pull.raiseError(MalformedMessageBodyFailure("Part not terminated properly"))
+            } else {
+              Pull.output1(Part[F](hdrs, l)) >> splitOrFinish(DoubleCRLFBytesN, r, limit).flatMap {
+                case (hdrStream, remaining) =>
+                  if (hdrStream == streamEmpty) { //Empty returned if it worked fine
+                    Pull.done
+                  } else {
+                    tailrecParts[F](b, hdrStream, remaining, expectedBytes, limit)
+                  }
+              }
+            }
         }
-    }
+      }
 
-  private def splitHalf[F[_]: Sync](
+  /** Split a stream in half based on `values`,
+    * but check if it is either double dash terminated (end of multipart).
+    * SplitOrFinish also tracks a header limit size
+    *
+    * If it is, return the empty stream. if it is not, split on the `values`
+    * and raise an error if we lack a match
+    */
+  //noinspection ScalaStyle
+  private def splitOrFinish[F[_]: Sync](
       values: Array[Byte],
-      stream: Stream[F, Byte]): Pull[F, Nothing, (Stream[F, Byte], Stream[F, Byte])] = {
+      stream: Stream[F, Byte],
+      limit: Int): SplitStream[F] = {
+
+    /** Check if a particular chunk a final chunk, that is,
+      * whether it's the boundary plus an extra "--", indicating it's
+      * the last boundary
+      */
+    def checkIfLast(c: Chunk[Byte], rest: Stream[F, Byte]): SplitStream[F] =
```

Comment:
I was conflicted on moving this out of the inner body if the method, but then I realized that this needs a reference to `go` for the continuation. It's a large function but mostly a side effect of us working on chunks directly so idk how to improve this :(